library/std/src/io/mod.rs

```diff
@@ -2462,6 +2462,34 @@ impl<R: Read> Iterator for Bytes<R> {
             };
         }
     }
+
+    default fn size_hint(&self) -> (usize, Option<usize>) {
+        (&self.inner as &dyn SizeHint).size_hint()
```

Comment:
The original implementation was much nicer, but wasn't allowed due to restrictions on specialization.

I am not sure what overhead the use of `dyn` would cause, but this should work to provide a better s`ize_hint`